design-proposals/kliup-29-explicit-keys.md

```diff
@@ -0,0 +1,154 @@
+**Status**: In Discussion | 
+**Discussion**: [Github PR](https://github.com/confluentinc/ksql/pull/5530)
+
+**tl;dr:** Up until now ksqlDB has added an implicit `ROWKEY STRING (PRIMARY) KEY` to a `CREATE TABLE`
+or `CREATE STREAM` statement that does not itself explicitly define a key column. This KLIP proposes
+removing this implicit column creation and instead requiring tables to define their PRIMARY KEY, and
+changing `CREATE STREAM` to create a stream without a KEY column, should not be defined.
+
+## Motivation and background
+
+Implicitly adding columns is confusing for users. They are left wondering where did this column come from? 
+For example:
+
+```
+ksal> CREATE STREAM S (ID INT, NAME STRING) WITH (...);
+Stream S Created.
+ksal> CREATE STREAM S2 AS SELECT ID, NAME FROM S;
+Key missing from projection. See https://cnfl.io/2LV7ouS.
+The query used to build `S2` must include the key column ROWKEY in its projection.
+```
+
+Understandably, the user may be left wondering where this `ROWKEY` column came from and why they need to
+add it to the projection.
+
+Now that ksqlDB supports more than just `STRING` key columns, it no longer makes sense to add an
+implicit `STRING` key column by default.  Better to let the user define _if_ there is a key column, and if
+so, what it's name and type are.
+
+There is no semantic need for a STREAM to have data in the key, and hence there should be no need for 
+there to be a KEY column defined.  This can also be useful when the data in the key is not in a format 
+ksqlDB can work with, e.g. an Avro, JSON or Protobuf key, or any other format.  It seems unwieldy and 
+confusing to add an implicit `STRING` key column if the actual key is empty or is not something ksqlDB 
+can read.
+
+Unlike STREAMs, TABLEs require a PRIMARY KEY column. Therefore, we should require the user to provide one.  
+
+## What is in scope
+
+* `CREATE TABLE` statements will require a `PRIMARY KEY` column to be defined.
+* `CREATE STREAM` statements without a `KEY` column defined will define a stream without a key column. 
+
+## What is not in scope
+
+* Everything else.
+
+## Value/Return
+
+Removal of implicitly added columns makes statements more declarative. They make it possible to describe 
+a stream with no data, or incompatible data, in the key. They make users _think_ about the name and type
+of the PRIMARY KEY column of their tables.
+
+Combined, these changes make the language easier to use and richer.
+
+## Public APIS
+
+### CREATE TABLE changes
+
+A `CREATE TABLE` statement that does not define a `PRIMARY KEY` column will now result in an error:
+
+```
+ksql> CREATE TABLE INPUT (A INT, B STRING) WITH (...);
+Tables require a PRIMARY KEY. Please define the PRIMARY KEY.
+```
+
+Statements with a `PRIMARY KEY` column will continue to work as before.
+
+Where schema inference is used, the error message will include an example of how to define a partial
+schema to add the primary key:
+
+```ql
```

Comment:
nit: `sql` (seems some code snippets above don't define `sql` as language? -- not sure if you want to add it?)

Reply:
Some have console style output, not pure SQL.